conversion/resource_generation.go

```diff
@@ -0,0 +1,413 @@
+					Id:           id,
+					IsSource:     false,
+					Region:       profile.DstConnectionProfile.Location,
+					BucketName:   "GCS-" + uuid.New().String(),
+					ValidateOnly: false},
+				ctx: ctx,
+			})
+		}
+	}
+	return sourceProfilesToCreate, dstProfilesToCreate, err
+}
+
+// 1. For each datashard, check if source and destination connection profile exists or not
+// 2. If source connection profile doesn't exists create it or validate if creation is possible.
+// 3. If validation is false and destination connection profile doesn't exists create a corresponding gcs bucket and then a destination connection profile
+func (r ResourceGenerationStruct) CreateResourcesForShardedMigration(ctx context.Context, projectId string, instanceName string, validateOnly bool, region string, sourceProfile profiles.SourceProfile, dsClient *datastream.Client) error {
+	var sourceProfilesToCreate []*ConnectionProfileReq
+	var dstProfilesToCreate []*ConnectionProfileReq
+
+	// Fetches list with resources which do not exist and need to be created
+	s := ResourceGenerationStruct{}
+	sourceProfilesToCreate, dstProfilesToCreate, err := s.getResourcesForCreation(ctx, projectId, sourceProfile, region, validateOnly, dsClient)
+	if err != nil {
+		return fmt.Errorf("resource generation failed %s", err)
+	}
+
+	// If validating resource creation, validate for all connection profiles. If creating, return error for the first resource creation that fails.
+	fastExit := false
+	if !validateOnly {
+		fastExit = true
+	}
+
+	var errorsList []error = []error{}
+
+	// Create or validate source connection profiles in parallel threads
+	resSourceProfiles, resCreationErr := common.RunParallelTasks(sourceProfilesToCreate, 20, r.prepareMinimalDowntimeResources, fastExit)
+	// If creation failed, perform cleanup of resources
+	if resCreationErr != nil && !validateOnly {
+		err = r.connectionProfileCleanUp(ctx, resourcesForCleanup)
+		if err != nil {
+			return fmt.Errorf("resource generation failed due to %s, resources created could not be cleaned up, please cleanup manually: %s", resCreationErr.Error(), err.Error())
+		} else {
+			return resCreationErr
+		}
+	} else if resCreationErr != nil {
+		return resCreationErr
+	}
+	for _, resource := range resSourceProfiles {
+		if resource.Result.Error != nil && validateOnly {
+			// If validation failed, append to list of errors
+			errorsList = append(errorsList, resource.Result.Error)
+		}
+	}
+
+	// Create destination connection profiles in parallel threads
+	if !validateOnly {
+		_, resCreationErr := common.RunParallelTasks(dstProfilesToCreate, 20, r.prepareMinimalDowntimeResources, fastExit)
+		if resCreationErr != nil {
+			err = r.connectionProfileCleanUp(ctx, resourcesForCleanup)
+			if err != nil {
+				return fmt.Errorf("resource generation failed due to %s, resources created could not be cleaned up, please cleanup manually: %s", resCreationErr.Error(), err.Error())
+			} else {
+				return resCreationErr
+			}
+		}
+	}
```

Comment:
This flow is a bit confusing to read tbh - `ValidateResource` calls into `CreateResource` with a boolean flag.

Shouldn't `ValidateResource` call into the underlying methods that `CreateResource` would use to create resources directly, and passing the validation flag to it directly, instead of hopping through `CreateResource`?

Reply:
This would enhance the readability of `CreateResource` since I shouldn't have to think about `validation` or not when reading that code.

Reply:
Perhaps a better naming convention should be used to make it less confusing. Let me change this